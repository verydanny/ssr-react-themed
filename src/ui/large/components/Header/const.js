export const MENU_LINKS = [
  {
    name: 'My Saved Properties',
    link: '/account/myrent/',
    tag: 'my_rent',
    hasBadge: true,
  },
  {
    name: 'Search',
    link: '/',
    tag: 'search',
  },
  {
    name: 'Moving Center',
    link: '/moving-center/',
    tag: 'moving_center',
  },
  {
    name: 'List a property',
    link: '/manage',
    tag: 'list_a_property',
  },
  {
    name: 'Blog',
    link: '/blog/',
    tag: 'blog',
  },
]
