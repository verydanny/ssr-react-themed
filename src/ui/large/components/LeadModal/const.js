export const LEAD_MODAL_ID = 'Modal/LeadModal'
export const THANK_YOU_MODAL_ID = 'Modal/ThankYouModal'
export const LEAD_MODAL_PROPS = [
  'listingId',
  'endecaId',
  'propertyLabel',
  'desktopPhone',
  'tplSource',
  'isPaid',
  'revenue',
]
