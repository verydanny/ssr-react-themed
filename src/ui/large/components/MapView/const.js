export const CONTACT_FOR_PRICE = 'Contact for Price'
