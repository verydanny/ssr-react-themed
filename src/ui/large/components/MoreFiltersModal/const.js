export const MORE_FILTERS_MODAL_ID = 'Modal/MoreFiltersModal'
