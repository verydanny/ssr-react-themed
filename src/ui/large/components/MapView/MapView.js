import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { themed } from 'react-themed'
import {
  Gmap,
  Layer,
} from '@rentpath/react-ui-core'
import env from 'config/env'
import {
  MapBounds,
  MapTracking,
  dataStyler,
  emitMapDragEnd,
  emitMapClick,
  emitDataClick,
  emitMapZoomChanged,
  MAP_ZOOM_CHANGED,
} from 'ui/shared/components/Map'
import ZoomInBanner from 'ui/shared/components/ZoomInBanner'
import { eventSubject } from 'ui/shared/components/Map/events'
import { ofType } from 'redux-observable'
import Markers from './Markers'

const MAP_CONTROLS = {
  fullscreenControl: false,
  mapTypeControl: true,
  mapTypeControlOptions: {
    position: 3, // google.maps.ControlPosition.TOP_RIGHT
  },
  streetViewControl: true,
  draggableCursor: 'auto',
  zoomControl: true,
}

const MAP_EVENTS = {
  onClick: emitMapClick,
  onDragEnd: emitMapDragEnd,
  onZoomChanged: emitMapZoomChanged,
}

const LAYER_EVENTS = {
  onClick: emitDataClick,
}

@themed(/^MapView/)
export default class MapView extends PureComponent {
  static propTypes = {
    center: PropTypes.object,
    clearListingData: PropTypes.func,
    geoShape: PropTypes.object,
    points: PropTypes.object,
    theme: PropTypes.object,
    searchByBoundingBox: PropTypes.func,
    mapInitComplete: PropTypes.func,
    selectedListingId: PropTypes.string,
    zoom: PropTypes.number,
    belowMinZoom: PropTypes.bool,
    currentMapZoom: PropTypes.number,
    mapZoomChanged: PropTypes.func.isRequired,
    count: PropTypes.number,
    total: PropTypes.number,
  }

  static defaultProps = {
    theme: {},
  }

  componentDidMount() {
    const { mapZoomChanged } = this.props

    eventSubject
      .pipe(ofType(MAP_ZOOM_CHANGED))
      .subscribe(() => {
        mapZoomChanged()
      })
  }

  get zoomInBanner() {
    if (this.props.belowMinZoom) {
      return (
        <ZoomInBanner
          text="Zoom in on the map to see more."
        />
      )
    }
    return null
  }

  render() {
    const {
      center,
      geoShape,
      searchByBoundingBox,
      selectedListingId,
      theme,
      zoom,
      points,
      mapInitComplete,
      belowMinZoom,
      currentMapZoom,
    } = this.props

    return (
      <Gmap
        className={theme.MapView}
        apiKey={env.GOOGLE_MAPS_API_KEY}
        mapControls={MAP_CONTROLS}
        center={center}
        zoom={zoom}
        stylingFunction={dataStyler}
        {...MAP_EVENTS}
      >
        {this.zoomInBanner}
        { !belowMinZoom &&
          <Markers geojson={points} selectedListingId={selectedListingId} />
        }
        <MapBounds
          boundingBoxHandler={searchByBoundingBox}
          currentMapZoom={currentMapZoom}
          mapClass={theme.MapView}
          onComplete={mapInitComplete}
        />
        { !!geoShape && <Layer geojson={geoShape} {...LAYER_EVENTS} /> }
        <MapTracking />
      </Gmap>
    )
  }
}
