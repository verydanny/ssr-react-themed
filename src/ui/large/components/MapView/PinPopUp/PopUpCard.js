import React, { Component } from 'react'
import PropTypes from 'prop-types'
// TODO: move to selector
import formatPhone from 'lib/utils/formatPhone'
import { themed } from 'react-themed-too'
import get from 'lodash/fp/get'
import pick from 'lodash/fp/pick'
import autobind from 'autobind-decorator'
import cn from 'classnames'
import { handlePdpClick } from 'lib/ui/listingActions'
import withInfo from 'ui/shared/components/withInfo'
import {
  LEAD_MODAL_ID,
  LEAD_MODAL_PROPS,
} from 'ui/large/components/LeadModal/const'
import { CONTACT_FOR_PRICE } from 'ui/large/components/MapView/const'

const SERVER = 'https://image.rent.com/'
const VIEW_SIZE = '74-62' // only used if view size is not found

@themed('*')
export class BasePopUpCard extends Component {
  static propTypes = {
    theme: PropTypes.object,
    tplSource: PropTypes.string,
    photos: PropTypes.array,
    listingId: PropTypes.string,
    priceText: PropTypes.string,
    openModal: PropTypes.func,
    showPhone: PropTypes.bool,
    desktopPhone: PropTypes.string,
    propertyLabel: PropTypes.string,
    listingSeoPath: PropTypes.string,
    onMouseEnter: PropTypes.func,
    onMouseLeave: PropTypes.func,
    propertyType: PropTypes.string,
    isCurrentSpotlight: PropTypes.bool,
    isActive: PropTypes.bool,
  }

  static defaultProps = {
    theme: {},
    photos: [],
  }

  constructor(props) {
    super(props)
    this.divRef = React.createRef()
    this.hasClickHandler = false
  }

  componentDidMount() {
    const node = this.divRef.current

    if (node) {
      this.hasClickHandler = true
      node.addEventListener('click', this.handlePopUpClick, false)
    }
  }

  componentDidUpdate() {
    const node = this.divRef.current

    if (node && !this.hasClickHandler) {
      this.hasClickHandler = true
      node.addEventListener('click', this.handlePopUpClick, false)
    }
  }

  componentWillUnmount() {
    const node = this.divRef.current

    if (node && this.hasClickHandler) {
      node.removeEventListener('click', this.handlePopUpClick, false)
    }
  }

  onContactPropertyClick() {
    const data = {
      page: 'srp_map',
      section: 'map_listing',
      item: 'check_availability_button',
    }
    this.createEventType('click', data)
  }

  onPhoneClick() {
    const data = {
      page: 'srp_map',
      section: 'map_listing',
      item: 'check_availability_phone',
    }
    this.createEventType('click', data)
  }

  get propertyLabel() {
    const { theme, propertyType, isCurrentSpotlight, showPhone, priceText } = this.props

    if (propertyType !== 'APARTMENTS' && !isCurrentSpotlight && priceText !== CONTACT_FOR_PRICE) {
      const withPhone = showPhone ? 'with' : 'without'
      return (
        <div
          className={cn(theme.Property_Rating, theme[`PopUpCard_PropertyLabel-${withPhone}Phone`])}
        >
          HOME FOR RENT
        </div>
      )
    }
    return (
      isCurrentSpotlight && (
        <div className={cn(theme['PopUpCard_PropertyLabel-withPhone'], theme.PopUpCard_SponsoredText)} >
          SPONSORED
        </div>
      )
    )
  }

  get renderPropertyLabel() {
    const { isCurrentSpotlight } = this.props

    if (isCurrentSpotlight) {
      return (
        this.sponsoredLabel
      )
    }
    return this.propertyLabel
  }

  photo(path) {
    // Get the size for this viewType, or if that is not defined use a fallback
    // TODO: LF how are we falling back for failed image requests?
    // i've kicked the question to product, should we be supporting
    // caption text for these photos as well? (if available)
    if (path) {
      const size = VIEW_SIZE
      return `${SERVER}${path}${size}`
    }
    return ''
  }

  @autobind
  handlePopUpClick(e) {
    // Determine if the click was within an A or Button element,
    // of within an element with a data-nopdplink attribute
    const nopdplink = e.target.closest('a,button,[data-nopdplink]')

    if (nopdplink) {
      const listingProps = pick(LEAD_MODAL_PROPS)(this.props)

      const { openModal } = this.props

      if (openModal) openModal(LEAD_MODAL_ID, listingProps)
      if (e.target.textContent === 'Contact Property') {
        this.onContactPropertyClick()
      } else {
        this.onPhoneClick()
      }
    } else {
      handlePdpClick(e, this.props.listingSeoPath)
    }
  }

  createEventType(type, data) {
    window.eventTracker.track(type, data)
  }

  render() {
    const {
      theme,
      photos,
      priceText,
      tplSource,
      listingId,
      propertyLabel,
      showPhone,
      onMouseEnter,
      onMouseLeave,
      desktopPhone,
      isActive,
    } = this.props

    if (!listingId) return null

    const listingPhoto = get('path')(photos[0])
    const missingPhotos = photos.length === 0

    return (
      <div
        className={cn(theme.PopUpFrame, showPhone && theme['PopUpFrame-withPhone'])}
        data-tag_item="padding_box"
        data-tag_section="map_listing"
        data-tag_listing_id={listingId}
        data-tag_tpl_source={tplSource}
        data-tid="popup_card"
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        role="presentation"
        ref={this.divRef}
      >
        <div
          className={cn(
            theme.PopUpImage,
            { [theme['PopUpImage-missing']]: missingPhotos }
          )}
          data-tag_item="image"
        >
          {!missingPhotos &&
            <img
              alt=""
              src={this.photo(listingPhoto)}
              key={this.photo(listingPhoto)}
            />
          }
        </div>
        <div className={theme.PopUpListingInfo}>
          <div className={theme.PopUpListingInfo_Header}>
            <div className={theme.PopUpListingInfo_Price} data-tid="popup-property-price" >
              {priceText}
            </div>
            {this.propertyLabel}
          </div>
          <div className={theme.PopUpListingInfo_PropertyName} data-tid="popup-property-name">
            {propertyLabel}
          </div>
          {isActive &&
            <div data-nopdplink className={theme.PopUpListingInfo_ContactRow} data-tid="popup-contact-row">
              <span
                className={theme.PopUpListingInfo_Contact}
                data-tag_item="check_availability_button"
              >
                Contact Property
              </span>
              {showPhone &&
                <span>
                  <span className={theme.PopUpListingInfo_Separator}>
                    &nbsp;â€¢&nbsp;
                  </span>
                  <span
                    className={theme.PopUpListingInfo_Contact}
                    data-tag_item="check_availability_phone"
                  >
                    {formatPhone(desktopPhone)}
                  </span>
                </span>
              }
            </div>
          }
        </div>
      </div>
    )
  }
}

export default withInfo(BasePopUpCard)
