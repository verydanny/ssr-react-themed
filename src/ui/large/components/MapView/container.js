import { connect } from 'react-redux'
import {
  mapInitComplete,
  searchByBoundingBox,
  mapZoomChanged,
} from 'app/store/shared/actions'
import {
  getCenter as getCenterSelector,
} from 'app/selectors/criteria'
import {
  getPoints,
  selectedListingId as selectedListingIdSelector,
  getTotal as getTotalSelector,
} from 'app/selectors/listings'
import {
  getPointsCount as getPointsCountSelector,
  getViewportZoom as getZoomSelector,
  getCurrentMapZoom as getCurrentMapZoomSelector,
  isBelowMinZoom as isBelowMinZoomSelector,
} from 'app/selectors/map'
import {
  getGeoShape as geoShapeSelector,
} from 'app/selectors/shape'
import MapView from './MapView'

const mapStateToProps = state => ({
  center: getCenterSelector(state),
  geoShape: geoShapeSelector(state),
  points: getPoints(state),
  selectedListingId: selectedListingIdSelector(state),
  zoom: getZoomSelector(state),
  belowMinZoom: isBelowMinZoomSelector(state),
  currentMapZoom: getCurrentMapZoomSelector(state),
  count: getPointsCountSelector(state),
  total: getTotalSelector(state),
})

const mapDispatchToProps = {
  searchByBoundingBox,
  mapInitComplete,
  mapZoomChanged,
}

export default connect(mapStateToProps, mapDispatchToProps)(MapView)
