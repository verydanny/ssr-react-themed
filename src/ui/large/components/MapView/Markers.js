import React, { Fragment, PureComponent } from 'react'
import PropTypes from 'prop-types'
import autobind from 'autobind-decorator'
import { Markers as MapMarkers } from '@rentpath/react-ui-core'
import {
  emitMarkerClick,
  emitMarkerMouseOver,
  emitMarkerMouseOut,
  markerNormalIcon,
} from 'ui/shared/components/Markers'
import PinPopUp from './PinPopUp'

const MARKER_EVENTS = {
  onClick: emitMarkerClick,
  onMouseOver: emitMarkerMouseOver,
  onMouseOut: emitMarkerMouseOut,
}

export default class Markers extends PureComponent {
  static propTypes = {
    map: PropTypes.object,
    geojson: PropTypes.object,
    selectedListingId: PropTypes.string, // eslint-disable-line react/no-unused-prop-types
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    const selectedMarker = prevState.markers[nextProps.selectedListingId]

    if (selectedMarker !== prevState.selectedMarker) {
      return { selectedMarker }
    }

    return null
  }

  state = {
    selectedMarker: null,
    markers: {},
  }

  @autobind
  updateMarkers(markers) {
    this.setState({ markers })
  }

  markerFromFeature({ properties }) {
    return {
      icon: markerNormalIcon(properties),
      isPaid: properties.isPaid,
      title: null,
      zIndex: properties.isActive ? 2 : 1,
      ...MARKER_EVENTS,
    }
  }

  render() {
    const { selectedMarker } = this.state
    const { map, geojson } = this.props

    return (
      <Fragment>
        <MapMarkers
          map={map}
          geojson={geojson}
          marker={this.markerFromFeature}
          onMarkersReady={this.updateMarkers}
        />
        <PinPopUp
          map={map}
          selectedMarker={selectedMarker}
        />
      </Fragment>
    )
  }
}
