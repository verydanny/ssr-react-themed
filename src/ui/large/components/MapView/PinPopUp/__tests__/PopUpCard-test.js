import React from 'react'
import { CONTACT_FOR_PRICE } from 'ui/large/components/MapView/const'
import { mount, shallow } from 'enzyme'
import { BasePopUpCard } from '../PopUpCard'

const PopUpCard = BasePopUpCard.WrappedComponent

const theme = {
  'PopUpImage-missing': 'PopUpImage-missing',
  PopUpImage: 'PopUpImage',
  noPriceText: 'PopUpCard_PropertyLabel-withPhone',
}

const listings = {
  houseWithPhotos: {
    listingId: '101',
    propertyType: 'HOUSE',
    isActive: true,
    priceText: '$1,800',
    propertyLabel: 'Foo Bar 123',
    showPhone: false,
    desktopPhone: '1231231234',
    photos: [
      { path: 'foobar' },
    ],
  },
  houseWithNoPrice: {
    listingId: '123456',
    propertyType: 'HOUSE',
    isActive: true,
    priceText: CONTACT_FOR_PRICE,
    propertyLabel: 'Foo Bar 123',
    showPhone: true,
    desktopPhone: '1231231234',
    photos: [
      { path: 'foobar' },
    ],
  },
  apartmentSpotlight: {
    listingId: '102',
    propertyType: 'APARTMENTS',
    propertyLabel: 'Foo Bar 123',
    showPhone: true,
    isActive: true,
    desktopPhone: '1231231234',
    aggregates: {
      prices: {
        low: 1250,
        high: 1800,
      },
    },
    isCurrentSpotlight: true,
  },
  apartmentWithoutPhotos: {
    listingId: '103',
    propertyType: 'APARTMENTS',
    aggregates: {
      prices: {
        low: 1800,
        high: 1800,
      },
    },
  },
  apartmentWithEmptyPhotos: {
    listingId: '103',
    propertyType: 'APARTMENTS',
    aggregates: {
      prices: {
        low: 1800,
        high: 1800,
      },
    },
    photos: [],
  },
  inactiveListing: {
    listingId: '104',
    isActive: false,
    priceText: 'Price Unavailable',
    desktopPhone: '1231231234',
    propertyType: 'APARTMENTS',
    aggregates: {
      prices: {
        low: 1800,
        high: 1800,
      },
    },
    photos: [],
  },
}

describe('PopUpCard', () => {
  describe('apartment snapshot', () => {
    it('renders', () => {
      const wrapper = mount(<PopUpCard {...listings.apartmentSpotlight} />)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
  describe('house snapshot', () => {
    it('renders', () => {
      const wrapper = mount(<PopUpCard {...listings.houseWithPhotos} />)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
  describe('inactive snapshot', () => {
    it('renders', () => {
      const wrapper = mount(<PopUpCard {...listings.inactiveListing} />)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})

describe('PopUpCard photo', () => {
  it('contains image missing class when no photos are in listing', () => {
    const wrapper = mount(<PopUpCard {...listings.apartmentWithoutPhotos} theme={theme} />)
    expect(wrapper.find('div.PopUpImage-missing').exists()).toEqual(true)
  })

  it('contains image missing class when photos are empty', () => {
    const wrapper = mount(<PopUpCard {...listings.apartmentWithEmptyPhotos} theme={theme} />)
    expect(wrapper.find('div.PopUpImage-missing').exists()).toEqual(true)
  })

  it('does not contain image missing class when photo is available', () => {
    const wrapper = mount(<PopUpCard {...listings.houseWithPhotos} theme={theme} />)
    expect(wrapper.find('div.PopUpImage-missing').exists()).toEqual(false)
  })

  it('does not contain child image when photo is not available', () => {
    const wrapper = mount(<PopUpCard {...listings.apartmentWithoutPhotos} theme={theme} />)
    expect(wrapper.find('div.PopUpImage-missing').children().length).toEqual(0)
  })
})

describe('Inactive PopupCard Info', () => {
  it('does not display Contact Property CTA and phone number', () => {
    const wrapper = shallow(<PopUpCard {...listings.inactiveListing} theme={theme} />)
    expect(wrapper.find('[data-tid="popup-contact-row"]')).toHaveLength(0)
  })
})

describe('No Price PopUpCard', () => {
  it('Does not show HOME FOR RENT label when priceText unavailable', () => {
    const wrapper = mount(<PopUpCard {...listings.houseWithNoPrice} theme={theme} />)
    expect(wrapper.find('HOME FOR RENT').exists()).toBe(false)
  })
})
