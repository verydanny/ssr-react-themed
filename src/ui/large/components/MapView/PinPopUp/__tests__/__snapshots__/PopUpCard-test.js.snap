// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopUpCard apartment snapshot renders 1`] = `

<div class
     data-tag_item="padding_box"
     data-tag_section="map_listing"
     data-tag_listing_id="102"
     data-tid="popup_card"
     role="presentation"
>
  <div class="undefined"
       data-tag_item="image"
  >
  </div>
  <div>
    <div>
      <div data-tid="popup-property-price">
      </div>
      <div class>
        SPONSORED
      </div>
    </div>
    <div data-tid="popup-property-name">
      Foo Bar 123
    </div>
    <div data-nopdplink="true"
         data-tid="popup-contact-row"
    >
      <span data-tag_item="check_availability_button">
        Contact Property
      </span>
      <span>
        <span>
          &nbsp;â€¢&nbsp;
        </span>
        <span data-tag_item="check_availability_phone">
          (123) 123-1234
        </span>
      </span>
    </div>
  </div>
</div>

`;

exports[`PopUpCard house snapshot renders 1`] = `

<div class
     data-tag_item="padding_box"
     data-tag_section="map_listing"
     data-tag_listing_id="101"
     data-tid="popup_card"
     role="presentation"
>
  <div class
       data-tag_item="image"
  >
    <img alt
         src="https://image.rent.com/foobar74-62"
    >
  </div>
  <div>
    <div>
      <div data-tid="popup-property-price">
        $1,800
      </div>
      <div class>
        HOME FOR RENT
      </div>
    </div>
    <div data-tid="popup-property-name">
      Foo Bar 123
    </div>
    <div data-nopdplink="true"
         data-tid="popup-contact-row"
    >
      <span data-tag_item="check_availability_button">
        Contact Property
      </span>
    </div>
  </div>
</div>

`;

exports[`PopUpCard inactive snapshot renders 1`] = `

<div class
     data-tag_item="padding_box"
     data-tag_section="map_listing"
     data-tag_listing_id="104"
     data-tid="popup_card"
     role="presentation"
>
  <div class="undefined"
       data-tag_item="image"
  >
  </div>
  <div>
    <div>
      <div data-tid="popup-property-price">
        Price Unavailable
      </div>
    </div>
    <div data-tid="popup-property-name">
    </div>
  </div>
</div>

`;
